website/src/pages/tech-specs.md

```diff
@@ -0,0 +1,371 @@
+| Partitioning key [Optional] | Value of this field defines the directory hierarchy within the table base path. This essentially provides an hierarchy isolation for managing data and related metadata |
+| Row key(s)                     | Record keys uniquely identify a record/row within each partition if partitioning is enabled |
+| Ordering field(s)           | Hudi guarantees the uniqueness constraint of row key and the conflict resolution configuration manages strategies on how to disambiguate when multiple records with the same keys are to be merged into the dataset. The resolution logic can be based on an ordering field or can be custom, specific to the dataset. To ensure consistent behaviour dealing with duplicate records, the resolution logic should be commutative and idempotent |
+
+**Hudi metadata fields**
+
+Hudi format stores the user fields along with the row merged along with transactional metadata fields. These fields are encoded in the data-file format and available in the table schema. 
+
+| Hudi meta-fields             | Description                                                  |
+| ---------------------------- | ------------------------------------------------------------ |
+| _hoodie_commit_time [string] | Every modification to a Hudi dataset creates an entry into the Transaction timeline. This entry is identified with the commit time. This field matches to the commit time of the instant in the timeline that created this record. More on how to populate this in Hudi transactions section below. |
+| _hoodie_record_key           | Unique record key identifying the row within the partition   |
+| _hoodie_partition_path       | Partition path under which the data is organized into        |
+| _hoodie_file_name            | The data file name this record belongs to                    |
+| _hoodie_is_deleted           | Tombstone field to denote the record key is deleted          |
+
+
+
+## Transaction Log (Timeline)
+
+Data consistency in Hudi is provided using Multi-version Concurrency Control (MVCC). Every transactional action on the Hudi table creates a new entry (instant) in the timeline. All transactional actions follows the state transition below
+
+* **requested** - Action is requested to start on the timeline. 
+* **inflight** - Action has started running and is currently in-flight
+* **completed** - Action has completed running
+
+All actions and the state transitions are registered with the timeline using an atomic put of special meta-file inside the  *.hoodie* directory. The requirement from the underlying storage system is to support an atomic-put and read-after-write consistency. The meta file structure is as follows
+
+```
+[Action timestamp].[Action type].[Action state] 
+```
+
+* **Action timestamp** - Monotonically increasing value to denote strict ordering of actions in the timeline. This could be provided by an external token provider or rely on the system epoch time at millisecond granularity. 
```

Comment:
Yeah I think the format can leave the specifics on how to generate this monotonically increasing value to the implementation. But the caveat on retry may be worth calling out - not sure where. This is more like things to watch out for in the implementation, but there are tons of things we can mention here. I am inclined to leave it as-is for now.

Reply:
The problem is that this things are unfortunately outside of our the control of implementation: 

- For the first one to be upheld the cluster(s) to run Hudi jobs have to have clocks synchronized, which is an operational aspect outside of Hudi's control
- For the second one it's mostly to exemplify what locks are actually needed and used for (that we're not holding locks for the duration of the whole operation, which will kill the performance, but only holding it while we operate on the timeline)

My point here was that i think we should call out (not necessarily very verbosely) since these are assumptions about the externalities (cluster setup, configs provided) that we can't control